Stripe/STPCustomerContext.m

```diff
@@ -0,0 +1,154 @@
+    static STPCustomerContext *sharedInstance;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [[self alloc] initWithKeyManager:nil];
+    });
+    return sharedInstance;
+}
+
+- (instancetype)initWithKeyProvider:(nonnull id<STPEphemeralKeyProvider>)keyProvider {
+    STPEphemeralKeyManager *keyManager = [[STPEphemeralKeyManager alloc] initWithKeyProvider:keyProvider
+                                                                                  apiVersion:[STPAPIClient apiVersion]];
+    return [self initWithKeyManager:keyManager];
+}
+
+- (instancetype)initWithKeyManager:(STPEphemeralKeyManager *)keyManager {
+    self = [self init];
+    if (self) {
+        _cachedCustomerMaxAge = DefaultCachedCustomerMaxAge;
+        _keyManager = keyManager;
+        if (keyManager) {
+            [self retrieveCustomer:nil];
+        }
+    }
+    return self;
+}
+
+- (void)setKeyProvider:(id<STPEphemeralKeyProvider>)keyProvider {
+    _keyManager = [[STPEphemeralKeyManager alloc] initWithKeyProvider:keyProvider
+                                                           apiVersion:[STPAPIClient apiVersion]];
+    [self retrieveCustomer:nil];
+}
+
+- (void)setCustomer:(STPCustomer *)customer {
+    _customer = customer;
+    _customerRetrievedDate = (customer) ? [NSDate date] : nil;
+}
+
+- (BOOL)shouldUseCachedCustomer {
+    if (!self.customer || !self.customerRetrievedDate) {
+        return NO;
+    }
+    NSDate *now = [NSDate date];
+    return [now timeIntervalSinceDate:self.customerRetrievedDate] < self.cachedCustomerMaxAge;
+}
+
+- (void)retrieveCustomer:(STPCustomerCompletionBlock)completion {
+    if ([self shouldUseCachedCustomer]) {
+        if (completion) {
+            completion(self.customer, nil);
+        }
+        return;
+    }
+    if (!self.keyManager) {
+        completion(nil, [NSError stp_customerContextMissingKeyProviderError]);
+        return;
+    }
+    [self.keyManager getCustomerKey:^(STPEphemeralKey *ephemeralKey, NSError *retrieveKeyError) {
+        if (retrieveKeyError) {
+            if (completion) {
+                completion(nil, retrieveKeyError);
+            }
+            return;
+        }
+        [STPAPIClient retrieveCustomerUsingKey:ephemeralKey completion:^(STPCustomer *customer, NSError *error) {
+            if (customer) {
+                self.customer = customer;
+            }
+            if (completion) {
+                completion(customer, error);
+            }
+        }];
+    }];
+}
+
+- (void)attachSourceToCustomer:(id<STPSourceProtocol>)source completion:(STPErrorBlock)completion {
+    if (!self.keyManager) {
+        completion([NSError stp_customerContextMissingKeyProviderError]);
+        return;
+    }
+    [self.keyManager getCustomerKey:^(STPEphemeralKey *ephemeralKey, NSError *retrieveKeyError) {
+        if (retrieveKeyError) {
+            if (completion) {
+                completion(retrieveKeyError);
```

Comment:
Ah yep, I think right now PaymentContext pretty consistently dispatches to main from `apiAdapter` calls, but it seems reasonable to make CustomerContext always dispatch to main.